src/pip/_internal/network/auth.py

```diff
@@ -46,40 +46,31 @@ class KeyRingCli:
     PATH.
     """
 
-    @staticmethod
-    def _quote(string: Optional[str]) -> str:
-        return f"'{string}'"
-
+    @classmethod
     def get_credential(
-        self, service_name: str, username: Optional[str]
+        cls, service_name: str, username: Optional[str]
     ) -> Optional[KeyRingCredential]:
-        cmd = ["keyring", "get", self._quote(service_name), self._quote(username)]
-        res = subprocess.run(cmd)
+        cmd = ["keyring", "get", service_name, str(username)]
+        res = subprocess.run(cmd, capture_output=True)
         if res.returncode:
             return None
-        return KeyRingCredential(username=username, password=res.stdout)
-
-    def set_password(self, service_name: str, username: str, password: str) -> None:
-        cmd = [
-            "echo",
-            self._quote(password),
-            "|",
-            "keyring",
-            "set",
-            self._quote(service_name),
-            self._quote(username),
-        ]
-        res = subprocess.run(cmd)
-        if res.returncode:
-            raise RuntimeError(res.stderr)
+        password = res.stdout.decode().strip("\n")
```

Comment:
Encoding issues possible here. On Windows, at least, it's not necessarily true that keywring will write its output in the same encoding as the pip process' default encoding.

Reply:
I think I've solved this using `PYTHONIOENCODING`